### DISCUSSION TITLE
where to find the list of attributes for `dag_run.logical_date`

### DISCUSSION THREAD

**Discussion by a-alwu** [apache/airflow#35835] [Q&A] [answered]
Hi all, I am new to Airflow and have been struggling to set up a workflow using the variable `dag_run.logical_date`. I have the following questions and hope you can help?

1. Where can I get the list of attributes for `dag_run.logical_date`? I looked through the official doc but cant find it. Saw somewhere it has attribute like `dag_run.logical_date.hour` , but i dont know what other attributes are, the type of the attributes...?
2. I want to specify a workflow to run current_hour - 2 , and have this expression "{{ dag_run.logical_date - macros.timedelta(hour=2) }}" but didnt work...Do you have any suggestions?
Please let me know. Thanks a lot.

**Comment by Taragolis** [comment]
> Where can I get the list of attributes for dag_run.logical_date

`dag_run.logical_date` - it is a `pendulum.DateTime` object, you could have a look available methods, attributes, properties in the official documentation: https://pendulum.eustace.io/ , in addition it should be also a subclass of [`datetime.datetime`](https://docs.python.org/3/library/datetime.html#datetime.datetime) so you could look other attributes in the Python docs

> I want to specify a workflow to run current_hour - 2 , and have this expression "{{ dag_run.logical_date - macros.timedelta(hour=2) }}" but didnt work...Do you have any suggestions?

Need a bit more examples. It is hard to say what you tried to achieve without actual code. Maybe you want to configure when DagRun should be started, in that case you need to create own Timetable:
- [Authoring and Scheduling: Timetables](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/timetable.html)
- [AIP-39 Richer scheduler_interval](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-39+Richer+scheduler_interval) 
- [How-to Guides: Customizing DAG Scheduling with Timetables](https://airflow.apache.org/docs/apache-airflow/stable/howto/timetable.html)
- [Astronomer: DAG scheduling and timetables in Airflow](https://docs.astronomer.io/learn/scheduling-in-airflow)

**Reply by a-alwu** [reply]
Thanks a lot Andrey for your swift response. 
> dag_run.logical_date - it is a pendulum.DateTime object, you could have a look available methods, attributes, properties in the official documentation: https://pendulum.eustace.io/ , in addition it should be also a subclass of [datetime.datetime](https://docs.python.org/3/library/datetime.html#datetime.datetime) so you could look other attributes in the Python 
Will check the documentation. cheers!
> Need a bit more examples.
Our use case is to process events from kafka which are stored in a partitioned table (by date and hour). Because there are some delays in data arrival, we dont want to process data that came in the the current hour because the partition would be incomplete. Instead we want to process the hour partition with current_time - 2. 
According to your description, does it mean Airflow does not have built in functionalities for setting specific hour for the job to start?

**Reply by Taragolis** [reply, marked as answer]
By default Airflow use interval based cron scheduling:
- [Administration and Deployment: Scheduler](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/scheduler.html)
- [Administration and Deployment: Timetables comparisons](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/timetable.html#timetables-comparisons)

That mean if your DAG define `schedule=@hourly` or something similar then it start in the end of the period, and logical date would be data_interval.start, e.g.:
- Logical Date (Execution date): 2023-11-24T**20**:00:00.000Z
- Date Interval Start: 2023-11-24T**20**:00:00.000Z
- Date Interval End: 2023-11-24T**21**:00:00.000Z
- Execution Started After: 2023-11-24T**21**:00:00.000Z

In this case logical date would be at least 1 hour behind of actual time, but according to your requirement you need to have 2 hours.
This could be achieved by different way to achieve it.

Option 1:
In your operator subtract 1 hour from the logical date

Option 2:
Before run your operator wait until required time happen, e.g. by use [`airflow.sensors.time_sensor.TimeSensor`](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/sensors/time_sensor/index.html#airflow.sensors.time_sensor.TimeSensor) sensor

Option 3:
Write custom timetable with custom logic around [DagRunInfo.run_after](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/timetables/base/index.html#airflow.timetables.base.DagRunInfo.run_after)

Another options maybe also exists, I just first which I remember. In general builtin functionality enough for most of the cases, other cases potentially could be covered by custom Timetables

**Reply by a-alwu** [reply]
Thanks a lot Andrey, option 1 worked when i used (execution_date - macros.timedelta(hours=2)).isoformat()[:16]... as the datetime format we use is with UTC. 

If I may share some feedback, perhaps is good to add
1. The note you mentioned above to the dag_run.logical_date description. on the Airflow main page template there is no explicit description about this class itself and what this class is built on top of.
<img width="1381" alt="Screenshot 2023-11-30 at 11 07 39" src="https://github.com/apache/airflow/assets/39926641/cf8ca799-7509-4cf5-8652-bed20b9adf98">

> it is a pendulum.DateTime object, you could have a look available methods, attributes, properties in the official documentation: https://pendulum.eustace.io/ , in addition it should be also a subclass of [datetime.datetime](https://docs.python.org/3/library/datetime.html#datetime.datetime) so you could look other attributes in the Python docs

3. To add a note what the default datetime format is for dag_run.logical_date, and for UTC need to chain with the method .isoformat(). 
4. Most of the examples on the templates are daily. If will be great if you can add examples for hourly scheduling, or specified a time range to the doc. 

My colleagues and I are new to Airflow and some of us are not familiar with Python. Will be really useful to provide more explicit descriptions on the doc so that we can troubleshoot much quicker. 

Thanks a lot for your help!

**Reply by potiuk** [reply]
> My colleagues and I are new to Airflow and some of us are not familiar with Python. Will be really useful to provide more explicit descriptions on the doc so that we can troubleshoot much quicker.

Of course and now that you got helped. - this is a fantastic opportunity for you and your team to contribute back to free software you use.

Airflow is created by > 2700 contributors and many of them contribute documentation updates to make Airflow better. Now that you know what could be improved you are likely the best person in the world to describe it in the way that other people like you would understand and follow. You probably know best where you looked for it, where you did not find it and how it shoudl be phrased so that others would find it easier.

And it is super-simple - just go to the doc page, click "Suggest a change on this page" (bottom right on every doc page) and you willl have Pull Request opened for you in the right file to update - so that you will be able to update the documentation and add some more information - right there - not even leaving the GitHub UI. It's as simple as creating an issue or commenting on it.

Can we count on your help with it @a-alwu ?

**Reply by a-alwu** [reply]
Hi Jarek yes for sure. sorry for the late response. I will see where I can add this part to the documentation :)

**Reply by a-alwu** [reply]
If it is possible can you please check if the example I added for dag_run.logical_date is correct?
